crates/turbo-tasks-memory/src/cell.rs

```diff
@@ -1,37 +1,291 @@
+            | Cell::Full(box FullCell::UpdatedValue { content, .. }) => content.clone(),
+        }
     }
 
     pub fn track_read(&mut self, reader: TaskId) {
-        self.dependent_tasks.insert(reader);
+        match self {
+            Cell::Empty => {}
+            &mut Cell::Recomputing {
+                ref mut event,
+                updates,
+            } => {
+                *self = Cell::Full(box FullCell::Recomputing {
+                    event: event.take(),
+                    updates,
+                    dependent_tasks: AutoSet::from([reader]),
+                });
+            }
+            Cell::Full(box FullCell::Recomputing {
+                dependent_tasks, ..
+            })
+            | Cell::InitialValue {
+                dependent_tasks, ..
+            }
+            | Cell::Full(box FullCell::UpdatedValue {
+                dependent_tasks, ..
+            }) => {
+                dependent_tasks.insert(reader);
+            }
+        }
     }
 
     pub fn assign(&mut self, content: CellContent, turbo_tasks: &dyn TurboTasksBackendApi) {
-        self.content = content;
-        self.updates += 1;
-        // notify
-        if !self.dependent_tasks.is_empty() {
-            turbo_tasks.schedule_notify_tasks_set(&self.dependent_tasks);
+        match self {
+            Cell::Empty => {
+                *self = Cell::InitialValue {
+                    content,
+                    dependent_tasks: AutoSet::new(),
+                };
+            }
+            &mut Cell::Recomputing {
+                ref mut event,
+                updates,
+            } => {
+                event.notify(usize::MAX);
+                if updates == 1 {
+                    *self = Cell::InitialValue {
+                        content,
+                        dependent_tasks: AutoSet::new(),
+                    };
+                } else {
+                    *self = Cell::Full(box FullCell::UpdatedValue {
+                        content,
+                        dependent_tasks: AutoSet::new(),
+                        updates,
+                    });
+                }
+            }
+            &mut Cell::Full(box ref mut cell @ FullCell::Recomputing { .. }) => {
+                let FullCell::Recomputing {
+                    ref mut event,
+                    updates,
+                    ref mut dependent_tasks,
+                } = *cell else {
+                    unreachable!()
+                };
+                event.notify(usize::MAX);
+                if updates == 1 {
+                    *self = Cell::InitialValue {
+                        content,
+                        dependent_tasks: take(dependent_tasks),
+                    };
+                } else {
+                    *cell = FullCell::UpdatedValue {
+                        content,
+                        dependent_tasks: take(dependent_tasks),
+                        updates,
+                    };
+                }
```

Comment:
Ditto, I can't find a path that would cause a `FullCell::Recomputing` to have an `updates` greater than 1.